ledger/sandbox-on-x/src/classic/scala/com/daml/platform/ledger/SandboxServer.scala

```diff
@@ -104,25 +117,65 @@ final class SandboxServer(
       .map(path => Future(Files.write(path, Seq(port.toString).asJava)).map(_ => ()))
       .getOrElse(Future.unit)
 
-  private def loadPackages(writeService: WriteService)(implicit
-      executionContext: ExecutionContext
-  ): AbstractResourceOwner[ResourceContext, List[Unit]] =
-    ResourceOwner.forFuture(() =>
-      Future.sequence(
-        config.damlPackages.map { file =>
-          val submissionId = Ref.SubmissionId.assertFromString(UUID.randomUUID().toString)
-          for {
-            dar <- Future.fromTry(DarParser.readArchiveFromFile(file).toTry)
-            _ <- writeService
-              .uploadPackages(submissionId, dar.all, None)(
-                LoggingContext.ForTesting,
-                NoOpTelemetryContext,
+  private def loadPackages(writeService: WriteService, indexService: IndexService)(implicit
+      executionContext: ExecutionContext,
+      system: ActorSystem,
+      loggingContext: LoggingContext,
+  ): AbstractResourceOwner[ResourceContext, Unit] =
+    ResourceOwner.forFuture(() => {
+      val packageUploadTimeout = 10.seconds
+
+      val packageSubmissionsTrackerMap = config.damlPackages.map { file =>
+        val uploadCompletionPromise = Promise[Unit]()
+        system.scheduler.scheduleOnce(packageUploadTimeout) {
+          // Package upload timeout, meant to allow fail-fast for testing
+          // TODO Remove once in-memory backend (e.g. H2) is deemed stable
+          uploadCompletionPromise.tryFailure(
+            new RuntimeException(s"Package upload timeout after $packageUploadTimeout")
+          )
+          ()
+        }
+        UUID.randomUUID().toString -> (file, uploadCompletionPromise)
+      }.toMap
```

Comment:
I think lines above are mixing up 2 separate concerns: creation of auto-resolvable promise after 10 sec and  creation of the map of promises. I think this can be split into separate code paths.

Reply:
Went for conciseness here. Will split